src/vs/workbench/parts/extensions/electron-browser/extensionsActions.ts

```diff
@@ -1857,6 +1889,111 @@ export class ConfigureWorkspaceFolderRecommendedExtensionsAction extends Abstrac
+	private static readonly AddedLabel = localize('addedToWorkspaceRecommendations', "Added to workspace recommendations");
+
+	private static readonly AddClass = 'extension-action add-to-workspace-recommendations';
+	private static readonly AddingClass = 'extension-action add-to-workspace-recommendations adding';
+	private static readonly AddedClass = 'extension-action add-to-workspace-recommendations added';
+
+	private _extension: IExtension;
+	get extension(): IExtension { return this._extension; }
+	set extension(extension: IExtension) { this._extension = extension; this.update(); }
+
+	private disposables: IDisposable[] = [];
+
+	constructor(
+		@IFileService fileService: IFileService,
+		@IWorkspaceContextService contextService: IWorkspaceContextService,
+		@IEditorService editorService: IEditorService,
+		@IJSONEditingService jsonEditingService: IJSONEditingService,
+		@ITextModelService textModelResolverService: ITextModelService,
+		@ICommandService private commandService: ICommandService,
+		@INotificationService private notificationService: INotificationService,
+	) {
+		super(
+			'extensions.add-to-workspace-recommendations',
+			AddToWorkspaceRecommendationsAction.AddLabel,
+			contextService,
+			fileService,
+			editorService,
+			jsonEditingService,
+			textModelResolverService
+		);
+		this.class = AddToWorkspaceRecommendationsAction.AddClass;
+		this.enabled = false;
+		this.disposables.push(this.contextService.onDidChangeWorkbenchState(() => this.update()));
+		this.update();
+	}
+
+	private update(): void {
+		if (
+			this.contextService.getWorkbenchState() === WorkbenchState.EMPTY ||
+			!this.extension ||
+			this.extension.state !== ExtensionState.Installed
+		) {
+			this.enabled = false;
+		} else {
+			const folders = this.contextService.getWorkspace().folders;
+			const configurationFiles = folders
+				.map(workspaceFolder => workspaceFolder.toResource(paths.join('.vscode', 'extensions.json')));
+			const getFoldersRecommendedExtensionsPromises = configurationFiles.map(configurationFile => {
+				return this.getFolderRecommendedExtensions(configurationFile);
+			});
+			Promise.all(getFoldersRecommendedExtensionsPromises).then(foldersRecommendationsArray => {
+				// Don't enable the button if the selected extension is
+				// recommended in ANY the folders of the current workspace
+				this.enabled = !foldersRecommendationsArray.some((recommendations: string[]) => {
+					return recommendations.some(r => r === this.extension.id);
+				});
+				if (this.class === AddToWorkspaceRecommendationsAction.AddingClass) {
+					this.label = this.enabled ? AddToWorkspaceRecommendationsAction.AddLabel : AddToWorkspaceRecommendationsAction.AddedLabel;
+					this.class = this.enabled ? AddToWorkspaceRecommendationsAction.AddClass : AddToWorkspaceRecommendationsAction.AddedClass;
+				}
+			});
+		}
+	}
+
+	run(): TPromise<any> {
+		if (!this.enabled) {
+			return null;
+		}
+
+		this.enabled = false;
```

Comment:
I thought setting this to false made the action disappear... Is that not the case?

Reply:
In the case of this action it will only disappear if `this.enabled` is `true` and the class is `AddClass`. If the class is either `AddingClass` or `AddedClass` then it'll be dimmed and unclickable instead (but still visible), which is what I'm doing in the next two lines. What really makes the button disappear is the combination of the `this.enabled` and `this.class` field. See the changes in the `extensionActions.css` 

```diff
 .monaco-action-bar .action-item.disabled .action-label.extension-action.enable,
 .monaco-action-bar .action-item.disabled .action-label.extension-action.reload,
+.monaco-action-bar .action-item.disabled .action-label.extension-action.add-to-workspace-recommendations:not(.adding):not(.added),
 .monaco-action-bar .action-item.disabled .action-label.disable-status.hide,
 .monaco-action-bar .action-item.disabled .action-label.malicious-status.not-malicious {
        display: none;
 }
```